### PR TITLE
update RelaxedR1CSGadget to work with FieldVar trait so we can plug in FpVar and NonNativeFieldVar indistinctly

### PR DESCRIPTION
**Pull request by arnaucube**
This is to be able to instantiate the CycleFoldCircuit over Curve2 constraint field, and check it's RelaxedR1CS relation non-natively inside the Curve1 constraint field, while reusing the same code that we already have for checking the main circuit RelaxedR1CS over Curve1 constraint field natively.

### CODE REVIEW

**Review by CPerezz — APPROVED** [review]
The PR looks good. Nothing to say about it.

IIUC, We just need now to link the two circuits (native and non-native parts) into a single CS which is what the REAL DECIDER will be doing. Is that right? Also we will need PI management and some glue. But so far looks like this is almost it.

I also would like to address the Decider interface (so that we can traitify it and make it more generic). 
As we're getting closer to the end-to-end test.
In any case, I agree is too early to abstract. So we might want to keep working like this until we have the HyperNova/ProtoGalaxy deciders. Which will bring some light on the way to go!

File: src/decider/circuit.rs [hunk 1]
```diff
@@ -15,12 +14,14 @@ use crate::Error;
 pub type ConstraintF<C> = <<C as CurveGroup>::BaseField as Field>::BasePrimeField;
 
 #[derive(Debug, Clone)]
-pub struct RelaxedR1CSGadget<F: PrimeField> {
+pub struct RelaxedR1CSGadget<F: PrimeField, CF: PrimeField, FV: FieldVar<F, CF>> {
```
[the remark this answers is not in the document]
Comment (arnaucube) [reply src/decider/circuit.rs:17]: ConstraintField, is the field used as the constraint system field.

File: src/decider/circuit.rs [hunk 2]
```diff
@@ -309,4 +361,41 @@ mod tests {
...
+    #[test]
+    fn test_relaxed_r1cs_nonnative_circuit() {
+        let cs = ConstraintSystem::<Fq>::new_ref();
+        // in practice we would use CycleFoldCircuit, but is a very big circuit (when computed
+        // non-natively inside the RelaxedR1CS circuit), so in order to have a short test we use a
+        // custom circuit.
+        let circuit = CustomTestCircuit::<Fq>::new(10);
+        circuit.generate_constraints(cs.clone()).unwrap();
+        cs.finalize();
+        dbg!(cs.num_constraints());
```
[the remark this answers is not in the document]
Comment (arnaucube) [reply src/decider/circuit.rs]: Right, removed the dbgs for now, and will prepare a separate branch for all the metrics of number of constraints (already have most of it in a wip branch).

File: src/folding/nova/cyclefold.rs [hunk 3]
```diff
@@ -386,6 +386,7 @@ where
         // the CycleFoldCircuit are the sames used in the public inputs 'x', which come from the
         // AugmentedFCircuit.
         // TODO: Issue to keep track of this: https://github.com/privacy-scaling-explorations/folding-schemes/issues/44
+        // TODO add also r_bits as public input, and check it in the main circuit
```
[the remark this answers is not in the document]
Comment (arnaucube) [reply src/folding/nova/cyclefold.rs]: Updated the code to point to a the new issue that tracks this: https://github.com/privacy-scaling-explorations/folding-schemes/issues/48

**Review by han0110 — APPROVED** [review]
LGTM!

**Pull request merged**